### PR TITLE
SMES and APC EMP proofing

### PR DESCRIPTION
**Pull request by soslow4u**
## About The Pull Request
request

image

this guy would save the day 

## Why It's Good For The Game

me dunno, lads asked for it

## Details

Three new stock parts for only purpose for EMP shielding. `/obj/item/stock_part/emp_shield` actually does nothing besides existing, but it tries its best yknow

Two recipes for APC and SMES shieldings

Placeholder sprite, woo

No mapedit support for now before maintainers would say their word about it

## Me wants

Pls give me a good sprite for it, probably two for each usable item

Bonus points if you do overlays for APC and SMES

## Changelog
:cl:
add: Two new stock parts to pimp up your APC and SMES against pesky EMPs. Don't forget to stick it in.
add: APC EMP shielding: requires 4 metal sheets, 2 uranium sheets, 10 pieces of wire and little bit of tool whacking
add: SMES EMP shielding: 10 metal sheets, 5 uranium sheets, full coil of wire and, you got it right, little bit of tool whacking
/:cl:

### CODE REVIEW

**Comment by Kurgis** [comment]
I suggest removing the SMES type, just make itba single stock part which can be generally applied to anything.

**Comment by soslow4u** [comment]
Is there real need to make every machine EMP-proof?

**Comment by Kurgis** [comment]
No, but I see no reason for Technomancers to need two different objects for the same purpose and function.

File: code/modules/power/apc.dm [hunk 1]
```diff
@@ -61,6 +61,11 @@
 /obj/machinery/power/apc/hyper/critical
 	is_critical = 1
 
+/obj/item/stock_part/emp_shield/apc
+	name = "APC EMP shield"
+	desc = "Device that would save your APC from electro-magnetic pulses. Probably. \nAttention! Manufacturer takes no responsibility for any harm caused by this device."
```
Comment (Hopekz) [inline code/modules/power/apc.dm:66]:
```suggestion
	desc = "Device that would save your APC from electromagnetic pulses. Probably. \nAttention! The manufacturer takes no responsibility for any harm caused by this device."
```

File: code/modules/power/smes_construction.dm [hunk 2]
```diff
@@ -50,6 +50,19 @@
...
+	name = "machinery emp shielding"
+	desc = "Weird metal box full of wires. This one looks salvaged."
+	icon = 'icons/obj/assemblies.dmi' //placeholder for now
+	icon_state = "armor-igniter-analyzer"
+	matter = list(MATERIAL_STEEL = 4)
+	origin_tech = list(TECH_MAGNET = 1, TECH_POWER = 2, TECH_MATERIAL = 2)
+
+/obj/item/stock_part/emp_shield/smes
+	name = "SMES EMP shield"
+	desc = "Device that would save your SMES from electro-magnetic pulses. Probably. \nAttention! Manufacturer takes no responsibility for any harm caused by this device."
```
Comment (Hopekz) [inline code/modules/power/smes_construction.dm:63]:
```suggestion
	desc = "Device that would save your SMES from electromagnetic pulses. Probably. \nAttention! The manufacturer takes no responsibility for any harm caused by this device."
```

**Review by genessee596 — APPROVED** [review]
this still needs code approval - i smell something fishy here - but i approve the idea

**Comment by soslow4u** [comment]
honestly I have idea just to ditch this pr for future technomancers' stability core thing

File: code/modules/power/apc.dm [hunk 3]
```diff
@@ -462,6 +468,16 @@
 
 		if(QUALITY_PRYING)
 			if(opened)
+				if (emp_shield)
+					if (prob(50) && electrocute_mob(usr, terminal.powernet, terminal))
+						return
+					if (!do_after(usr, 20 SECONDS, src)) // More coils = takes longer to disassemble. It's complex so largest one with 5 coils will take 50s
```
Comment (drexample) [inline code/modules/power/apc.dm:474]: Why !do_after with a return instead of if(do_after) and what comes afterwards?

**Review by drexample — CHANGES_REQUESTED** [review]

File: code/modules/power/apc.dm (hunk 3 above)
Comment (TorinoFermic) [reply code/modules/power/apc.dm:474]: Also the comment line does not match with what the duration states

**Comment by drexample** [comment]
@soslow4u Are ya workin', son?

File: code/modules/power/apc.dm (hunk 3 above)
Comment (soslow4u) [reply code/modules/power/apc.dm:474]:
```
if(!do_after)
    return
do_something
```
is the same as
```
if(do_after)
    do_something
```

Oh, and about comment - this block was copy-pasted from SMES coils insertions, didn't pay attention to comments

**Review by genessee596 — APPROVED** [review]

**Comment by TorinoFermic** [comment]
Should we get a proper sprite for it ?

**Review by Evankhell561 — CHANGES_REQUESTED** [review]
my main problem with your pr is the path you have chosen for the new item.
`/obj/item/stock_part/`

**Comment by Kurgis** [comment]
> Should we get a proper sprite for it ?
Personally I find the sprite to be irrelevant, it will only stall finished code PRs with spriting restraints before they get closed and forgotten. Not to say sprite isn't important, but it can be done anytime later.

**Comment by soslow4u** [comment]
> 
> 
> my main problem with your pr is the path you have chosen for the new item.
> `/obj/item/stock_part/`

Can you explain the issue?

**Comment by Evankhell561** [comment]
> Can you explain the issue?

 It is complicated to explain but I will try:
1. This does not behave like a stock part at all.
2. the variable rating does not matter, it only matters if the object is present or not.
3. does not affect the proc stock part.
4. rped acts in strange and different ways than other stock parts when it comes to this item.
In short, this new item is not a stock part and should not have that path.

File: code/datums/craft/recipes/misc.dm [hunk 4]
```diff
@@ -249,7 +249,7 @@
 	)
 	variation_type = CRAFT_VARIATION
 
-/datum/craft_recipe/pipe
+`/datum/craft_recipe/pipe
```
Comment (Evankhell561) [inline code/datums/craft/recipes/misc.dm:252]: ??

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

**Comment by genessee596** [comment]
This PR will be automatically closed in 2 days if these issues aren't even attempted to be fixed.

File: code/modules/power/apc.dm [hunk 5]
```diff
@@ -561,6 +578,23 @@
...
+			if (prob(50) && electrocute_mob(usr, terminal.powernet, terminal))
+				return
+			if (!do_after(user, 30 SECONDS, src))
+				return
+
+		to_chat(usr, SPAN_NOTICE("You install the EMP shield into the [src]."))
+		user.drop_item()
+		emp_shield = I
+		emp_shield.loc = src
+		RefreshParts()
```
Comment (Evankhell561) [inline code/modules/power/apc.dm:595]: why?

File: code/modules/power/smes_construction.dm [hunk 6]
```diff
@@ -356,6 +385,24 @@
...
+				if (!do_after(user, 100 * cur_coils, src))
+					return
+
+				to_chat(usr, SPAN_NOTICE("You install the EMP shield into the SMES unit."))
+				user.drop_item()
+				emp_shield = W
+				emp_shield.loc = src
+				component_parts += W
+				
+				RefreshParts()
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:405]: why?

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

**Review by Evankhell561 — CHANGES_REQUESTED** [review]
remember to add the parent item to the blacklist.

**Comment by soslow4u** [comment]
> remember to add the parent item to the blacklist.

I totally wouldn't, its ready to be used in random generation. Contains some mats and yields some levels for Moebius research

**Comment by Evankhell561** [comment]
> > remember to add the parent item to the blacklist.
> 
> I totally wouldn't, its ready to be used in random generation. Contains some mats and yields some levels for Moebius research


oh, there is no parent item for both objects? that makes me have to do another code quality check

**Comment by soslow4u** [comment]
> 
> 
> > > remember to add the parent item to the blacklist.
> > 
> > 
> > I totally wouldn't, its ready to be used in random generation. Contains some mats and yields some levels for Moebius research
> 
> oh, there is no parent item for both objects? that makes me have to do another code quality check

There is

**Comment by Evankhell561** [comment]
image

it does exist but it doesn't do anything at least give it a name that says it's broken, or something.

**Comment by soslow4u** [comment]
👀 
image

**Comment by Evankhell561** [comment]
by the way, all emp shield you have created should go to a unique dm for them.

File: code/modules/power/smes_construction.dm [hunk 7]
```diff
@@ -353,9 +382,23 @@
 				user.drop_item()
 				component_parts += W
 				W.loc = src
-				RefreshParts()
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:356]: ??

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/power/apc.dm [hunk 8]
```diff
@@ -561,6 +578,23 @@
...
+		else
+			if (prob(50) && electrocute_mob(usr, terminal.powernet, terminal))
+				return
+			if (!do_after(user, 30 SECONDS, src))
+				return
+
+		to_chat(usr, SPAN_NOTICE("You install the EMP shield into the [src]."))
+		user.drop_item()
+		emp_shield = I
+		emp_shield.loc = src
```
Comment (Evankhell561) [inline code/modules/power/apc.dm:594]: use the proc forceMove()

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/power/smes_construction.dm [hunk 9]
```diff
@@ -353,9 +382,23 @@
...
+				if (failure_probability && prob(failure_probability))
+					total_system_failure(failure_probability, user)
+					return
+				if (!do_after(user, 100 * cur_coils, src))
+					return
+
+				to_chat(usr, SPAN_NOTICE("You install the EMP shield into the SMES unit."))
+				user.drop_item()
+				emp_shield = W
+				emp_shield.loc = src
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:401]: ditto.

File: code/modules/power/smes_construction.dm (hunk 7 above)
Comment (soslow4u) [reply code/modules/power/smes_construction.dm:356]: You looking at outdated commit

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/power/apc.dm (hunk 8 above)
Comment (soslow4u) [reply code/modules/power/apc.dm:594]: Sure thing

File: code/modules/power/smes_construction.dm [hunk 10]
```diff
@@ -353,9 +382,23 @@
...
 				to_chat(usr, SPAN_WARNING("You can't insert more coils to this SMES unit!"))
+		else if (istype(W, /obj/item/emp_shield/smes))
+			if(emp_shield)
+				to_chat(user, SPAN_NOTICE("You can't add more EMP shielding to this SMES unit!"))
+				return
+			else
+				if (failure_probability && prob(failure_probability))
+					total_system_failure(failure_probability, user)
+					return
+				if (!do_after(user, 100 * cur_coils, src))
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:395]:
```suggestion
				if(!do_after(user, 100 * cur_coils, src))
```

File: code/modules/power/smes_construction.dm [hunk 11]
```diff
@@ -353,9 +382,23 @@
...
 				W.loc = src
-				RefreshParts()
 			else
 				to_chat(usr, SPAN_WARNING("You can't insert more coils to this SMES unit!"))
+		else if (istype(W, /obj/item/emp_shield/smes))
+			if(emp_shield)
+				to_chat(user, SPAN_NOTICE("You can't add more EMP shielding to this SMES unit!"))
+				return
+			else
+				if (failure_probability && prob(failure_probability))
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:392]:
```suggestion
				if(failure_probability && prob(failure_probability))
```

File: code/modules/power/smes_construction.dm [hunk 12]
```diff
@@ -353,9 +382,23 @@
 				user.drop_item()
 				component_parts += W
 				W.loc = src
-				RefreshParts()
 			else
 				to_chat(usr, SPAN_WARNING("You can't insert more coils to this SMES unit!"))
+		else if (istype(W, /obj/item/emp_shield/smes))
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:387]:
```suggestion
		else if(istype(W, /obj/item/emp_shield/smes))
```

**Comment by soslow4u** [comment]
> by the way, all emp shield you have created should go to a unique dm for them.

I'm not an artist myself and nobody contacted me about sprite so there is it

File: code/modules/power/apc.dm [hunk 13]
```diff
@@ -462,6 +468,17 @@
 
 		if(QUALITY_PRYING)
 			if(opened)
+				if (emp_shield)
```
Comment (Evankhell561) [inline code/modules/power/apc.dm:471]:
```suggestion
				if(emp_shield)
```

File: code/modules/power/apc.dm [hunk 14]
```diff
@@ -462,6 +468,17 @@
...
+					if (prob(50) && electrocute_mob(usr, terminal.powernet, terminal))
```
Comment (Evankhell561) [inline code/modules/power/apc.dm:472]:
```suggestion
					if(prob(50) && electrocute_mob(usr, terminal.powernet, terminal))
```

File: code/modules/power/apc.dm [hunk 15]
```diff
@@ -462,6 +468,17 @@
...
+						return
+					if (!do_after(usr, 10 SECONDS, src))
```
Comment (Evankhell561) [inline code/modules/power/apc.dm:474]:
```suggestion
					if(!do_after(usr, 10 SECONDS, src))
```

File: code/modules/power/apc.dm [hunk 16]
```diff
@@ -561,6 +578,23 @@
 
 		if(ABORT_CHECK)
 			return
+	if (istype(I, /obj/item/emp_shield/apc) && opened && terminal != null)
```
Comment (Evankhell561) [inline code/modules/power/apc.dm:581]:
```suggestion
	if(istype(I, /obj/item/emp_shield/apc) && opened && terminal != null)
```

File: code/modules/power/apc.dm [hunk 17]
```diff
@@ -561,6 +578,23 @@
...
+		if(emp_shield)
+			to_chat(user, SPAN_NOTICE("You can't add more EMP shielding to this APC!"))
+			return
+		else
+			if (prob(50) && electrocute_mob(usr, terminal.powernet, terminal))
```
Comment (Evankhell561) [inline code/modules/power/apc.dm:586]:
```suggestion
			if(prob(50) && electrocute_mob(usr, terminal.powernet, terminal))
```

File: code/modules/power/apc.dm [hunk 18]
```diff
@@ -561,6 +578,23 @@
...
+				return
+			if (!do_after(user, 30 SECONDS, src))
```
Comment (Evankhell561) [inline code/modules/power/apc.dm:588]:
```suggestion
			if(!do_after(user, 30 SECONDS, src))
```

File: code/modules/power/smes_construction.dm [hunk 19]
```diff
@@ -319,6 +332,22 @@
 
 		// Crowbar - Disassemble the SMES.
 		if(tool_type == QUALITY_PRYING)
+
+			if (emp_shield)
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:336]:
```suggestion
			if(emp_shield)
```

File: code/modules/power/smes_construction.dm [hunk 20]
```diff
@@ -319,6 +332,22 @@
...
+				playsound(get_turf(src), 'sound/items/Crowbar.ogg', 50, 1)
+				to_chat(user, SPAN_WARNING("You begin to yank the [emp_shield] out!"))
+				if (!do_after(usr, 100 * cur_coils, src)) // More coils = takes longer to disassemble. It's complex so largest one with 5 coils will take 50s
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:339]:
```suggestion
				if(!do_after(usr, 100 * cur_coils, src)) // More coils = takes longer to disassemble. It's complex so largest one with 5 coils will take 50s
```

File: code/modules/power/smes_construction.dm [hunk 21]
```diff
@@ -319,6 +332,22 @@
...
+					return
+
+				if (failure_probability && prob(failure_probability))
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:342]:
```suggestion
				if(failure_probability && prob(failure_probability))
```

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/power/smes_construction.dm (hunk 7 above)
Comment (Evankhell561) [reply code/modules/power/smes_construction.dm:356]: eh, no. This is a real change.

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

File: code/modules/power/smes_construction.dm [hunk 22]
```diff
@@ -343,19 +372,33 @@
...
 
-				if (failure_probability && prob(failure_probability))
+				if(failure_probability && prob(failure_probability))
 					total_system_failure(failure_probability, user)
 					return
 
 				to_chat(usr, SPAN_NOTICE("You install the coil into the SMES unit."))
 				user.drop_item()
 				component_parts += W
 				W.loc = src
```
Comment (Evankhell561) [inline code/modules/power/smes_construction.dm:384]:
```suggestion
				W.loc = src
				RefreshParts()
```

**Review by Evankhell561 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**